### ISSUE TITLE
PcapNetworkInterface does not respect the equals() and hashCode() contract

### ISSUE THREAD

**Issue by OLibutzki** [kaitoy/pcap4j#73]
The hashCode() method does not respect the equals() and hashCode() contract. 
Example: http://www.programcreek.com/2011/07/java-equals-and-hashcode-contract/

**Comment by kaitoy** [comment]
Thanks. I fixed it.
I don't remember why I wrote such silly code.

**Comment by OLibutzki** [comment]
Thanks for fixing it that quickly. Any plans for a release?

**Comment by kaitoy** [comment]
I will release the next version this weekend.

**Comment by OLibutzki** [comment]
In the previuos version of the class you just took the _name_ and the _local_ attribute into account for checking if a PcapNetworkInterface is equal. With b18162eca13500c6fb88cf6c5e6bbdc4046aecef you take all the attributes into account.

I wonder if this is the desired behaviour as in my opinion two instances are equal if the _name_ (and maybe the _local_ attribute) is/are equal.

In order to find a solution I had to wrap the PcapNetworkInterface class wghich is a dirty solution: https://github.com/openhab/openhab2-addons/blob/master/addons/binding/org.openhab.binding.amazondashbutton/src/main/java/org/openhab/binding/amazondashbutton/internal/pcap/PcapNetworkInterfaceWrapper.java

If you don't want to adjust the equals and hahsCode methods it would be very helpful, if PcapNetworkInterface wouldn't be final, so I do not have to wrap the class, but I can extend it and override the methods the way it fits for my use case.

Thanks a lot!

**Comment by kaitoy** [comment]
I intentionally changed it so.
PcapNetworkInterface represents an interface at a certain point.
For example, let's think a case where a PcapNetworkInterface instance is created as eth0, then the interface is replaced with another one that still corresponds to eth0, and then another PcapNetworkInterface instance is created. In this case, the first instance should not equal to the second one. My fix `b18162e` was for such cases.

The reason why PcapNetworkInterface is final is that I believe concrete classes should not be extended by inheritance because it's against the oop principal, destroys encapsulation, and exposes internal implementation to the subclasses. So, I don't want to remove final from PcapNetworkInterface.

The best practice to extend a concrete class is decoration, which is what you did in PcapNetworkInterfaceWrapper. It's really not dirty.

If you are interested in only interface names, it would probably be better to get names from PcapNetworkInterface objects and treat the strings only.
You can instantiate PcapHandle by interface names using [PcapHandle.Builder](https://github.com/kaitoy/pcap4j/blob/v1/pcap4j-core/src/main/java/org/pcap4j/core/PcapHandle.java#L1526)

**Comment by OLibutzki** [comment]
Thanks for answering.

> PcapNetworkInterface represents an interface at a certain point.

Ok, I think this is the main aspect. I built a service which calls Pcaps#findAllDevs() every 2 seconds in order to determine if a network interface has been removed or added. It would have been easier to put all PcapNetworkInterface instances into a HashMap or a Set and check if the instance already exists (by using the hashCode method).

Is there a notification mechanism in Pcap4J or do I have to poll for added or removed network interfaces the way I do it right now?

> The reason why PcapNetworkInterface is final is that I believe concrete classes should not be extended by inheritance because it's against the oop principal, destroys encapsulation, and exposes internal implementation to the subclasses.

I completely understand your concerns and motivation. I prefer composition over inheritance as well.
